Review request for Microsoft.DBforPostgreSQL to add version preview/2025-06-01-preview

mattboentoro opened the pull request:
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

![spec_pr_review_workflow_diagram](https://github.com/Azure/azure-rest-api-specs/assets/4429827/5bb5e7ce-8aff-4dbb-a3f8-0d9b68fef5b1)

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [x] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been created in adherence to [OpenAPI specs PR creation guidance](https://aka.ms/azsdkdocs/createopenapispec)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Convert existing [OpenAPI spec to TypeSpec spec](https://aka.ms/typespec/conversion) (do not combine this with implementing changes for a new API version).
  - [x] Other, please clarify:
    - Reorganizes file structure.
    - Renames files to make more sense of their contents.
    - Changes operationIds so that operation groups are nouns and operation names are verbs.
    - Corrects some minLength, maxLength, or pattern of definitions that were incorrect and didn't match the validations imposed on the backend.
    - Removes sensitive information that was slipped in some example files (like subscription identifiers, resource group names, or server names). Now only approved fictitious names are used.
    - Makes the style and grammr of all descriptions consistent across the board.
    - Adds some important missing examples.
    - Changes response codes to match those actually returned by our APIs, rather than trying to satisfy the validation rules checked in the build process.
    - Fixes multiple previously existing errors which had been supressed, but could be fixed in the swaggers.

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [x] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [x] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://aka.ms/azurerpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.
- [x] A [release plan](https://aka.ms/azsdkdocs/release-plans) has been created. If not, please create one as it will help guide you through the REST API and SDK creation process. 

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- If you don't have permissions to remove or add labels to the PR, request `write access` per [aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories](https://aka.ms/azsdk/access#request-access-to-rest-api-or-sdk-repositories)
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with ARM review (PR workflow diagram Step 2), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.
- For guidance on SDK breaking change review, refer to https://aka.ms/ci-fix.

Commit:
initial commit

Commit:
add tuning index and database changes

Commit:
remove tuning options change

Commit:
changes for default db name

mikekistler commented:
Breaking changes previously reviewed and approved in #33028

Commit:
add current state for index tuning

nasc17 commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2025-06-01-preview/TuningOptions.json:
```diff
@@ -0,0 +1,490 @@
+        "tags": [
+          "TuningOptions"
+        ],
+        "operationId": "TuningIndex_ListRecommendations",
```
Update all references that say TuningIndex to say TuningOptions, including example file names

mattboentoro replied:
addressed

Commit:
changed instance of tuning index name to tuning options

Commit:
update example reference from tuning index to tuning options

Commit:
change MVU and capability API to use PG17

ramoka178 commented:
1. Make the first commit to be the exact copy of old apiversion files , but in the new folder of the new apiversion.
2. From the second commit, make the changes you need on these copied files, starting from updating the apiversion string and any other new APIs etc.
3. Please fix all the failing required checks and then place it back in the queue.

mllorente-msft reviewed (approved):
The changes for the Table recommendations in TuningOptions LGTM. Looks like we are not changing the name of indexrecommendation object.

Commit:
change examples to use PG17

Commit:
run prettier

Commit:
fix prettier

Commit:
address model validation

nasc17 commented on specification/postgresql/resource-manager/Microsoft.DBforPostgreSQL/preview/2025-06-01-preview/examples/ClusterCreate.json:
```diff
@@ -13,7 +13,7 @@
         "administratorLogin": "examplelogin",
         "administratorLoginPassword": "examplepassword",
-        "version": "16",
+        "version": "17",
```
In portal we get "PostgreSQL version 17 is currently not supported for elastic cluster". Can you confirm if this is allowed now?

Commit:
revert cluster create example to use PG 16

mllorente-msft commented:
> > The changes for the Table recommendations in TuningOptions LGTM. Looks like we are not changing the name of indexrecommendation object.
> 
> @mllorente-msft should this be updated to fit more generalized names?

I think so. But I do not have clarity from any PMs in terms of naming. However, in my merged PR [here](https://msdata.visualstudio.com/Database%20Systems/_git/orcasql-breadth/pullrequest/1643923?path=/src/OrcasBreadthRP/PostgresArm/V20250601/Controllers/TuningOptions/IndexRecommendationController.cs&version=GBElasticServerV2&line=40&lineEnd=40&lineStartColumn=4&lineEndColumn=79&type=2&lineStyle=plain&_a=files&iteration=44&base=0) the resource name was changed to TuningRecommendedObjects. Perhaps ObjectRecommendation would be a better fit.

nasc17 commented:
> > > The changes for the Table recommendations in TuningOptions LGTM. Looks like we are not changing the name of indexrecommendation object.
> > 
> > 
> > @mllorente-msft should this be updated to fit more generalized names?
> 
> I think so. But I do not have clarity from any PMs in terms of naming. However, in my merged PR [here](https://msdata.visualstudio.com/Database%20Systems/_git/orcasql-breadth/pullrequest/1643923?path=/src/OrcasBreadthRP/PostgresArm/V20250601/Controllers/TuningOptions/IndexRecommendationController.cs&version=GBElasticServerV2&line=40&lineEnd=40&lineStartColumn=4&lineEndColumn=79&type=2&lineStyle=plain&_a=files&iteration=44&base=0) the resource name was changed to TuningRecommendedObjects. Perhaps ObjectRecommendation would be a better fit.

@mattboentoro can you also make sure the resources take out reference to "Index" as well please

Commit:
change the naming from IndexRecommendation* -> ObjectRecommendation*

Commit:
Auto update CMK key version

Commit:
change in the descriptions for the APIs in administrators, backups, capabilities, checknameavailability, configuration

nachoalonsoportillo committed:
update operation IDs for flexible server API to reflect create/update and list by subscription

nachoalonsoportillo committed:
refactor: update administrator examples and rename files for consistency

nachoalonsoportillo committed:
feat: add backups management API with create, delete, get, and list functionalities

nachoalonsoportillo committed:
Refactor code structure for improved readability and maintainability

nachoalonsoportillo committed:
refactor: remove deprecated CheckNameAvailability API and examples

nachoalonsoportillo committed:
Add example for updating PostgreSQL configuration 'constraint_exclusion'

- Introduced a new JSON example for the 'ConfigurationsUpdate' operation in the PostgreSQL resource manager.
- The example demonstrates how to update the 'constraint_exclusion' configuration with a user-defined value.
- Included detailed response structures for successful and asynchronous operations.

nachoalonsoportillo committed:
feat: update database examples and restructure for consistency

nachoalonsoportillo committed:
feat: add examples for firewall rules create, delete, get, and list operations

nachoalonsoportillo committed:
Update PostgreSQL examples to reflect new dates for restore and creation times

- Changed earliestRestoreDate and createdAt timestamps from "2025-01-01" to "2025-06-01" in multiple JSON example files.
- Updated pointInTimeUTC for various server creation and restoration examples to "2025-06-01".
- Adjusted initialRecommendedTime and lastRecommendedTime in tuning options examples to "2025-06-01".
- Modified startTime in virtual endpoint operations to "2025-06-01".

nachoalonsoportillo committed:
Add examples for PostgreSQL server management operations

- Introduced examples for promoting replicas as planned standalone and switchover.
- Added examples for restarting servers, including with failover options.
- Included examples for starting and stopping servers.
- Created examples for updating servers with various configurations, including custom maintenance windows, data encryption, major version upgrades, and Microsoft Entra integration.

nachoalonsoportillo committed:
Add migration examples for PostgreSQL resource manager

- Introduced new examples for creating migrations with various configurations:
  - MigrationsCreateWithOtherUsers.json
  - MigrationsCreateWithPrivateEndpointServers.json
  - MigrationsCreateWithRoles.json
- Added examples for migration operations:
  - MigrationsDelete.json
  - MigrationsGet.json
  - MigrationsGetMigrationWithSuccessfulValidationAndMigration.json
  - MigrationsGetMigrationWithSuccessfulValidationButMigrationFailure.json
  - MigrationsGetMigrationWithSuccessfulValidationOnly.json
  - MigrationsGetMigrationWithValidationFailures.json
  - MigrationsListByTargetServer.json
  - MigrationsUpdate.json

nachoalonsoportillo committed:
Remove deprecated Long Term Retention backup examples and add new API specifications for Backups Long Term Retention operations.

nachoalonsoportillo committed:
refactor: update PostgreSQL API specifications and examples for private endpoint connections

nachoalonsoportillo committed:
refactor: standardize response descriptions across PostgreSQL API specifications

nachoalonsoportillo committed:
Remove deprecated example JSON files and add new API specifications for PostgreSQL management, including captured logs and virtual network subnet usage. Update tuning options examples to reflect the latest API version and structure.

nachoalonsoportillo committed:
fix: update firewall rule name pattern to enforce valid naming conventions

nachoalonsoportillo committed:
feat: add new API specifications and examples for PostgreSQL administrators with Microsoft Entra integration

nachoalonsoportillo committed:
fix: update nextLink properties to use 'format: uri' for consistency across PostgreSQL API specifications

nachoalonsoportillo committed:
feat: add Advanced Threat Protection settings API specifications and examples for PostgreSQL management

nachoalonsoportillo committed:
fix: correct description casing for PrivateEndpoint and Database properties

nachoalonsoportillo committed:
fix: correct typo in 'format' property for nextLink in AdvancedThreatProtectionSettings.json

nachoalonsoportillo committed:
fix: update input files and suppressions for flexible server API specifications in readme.md

nachoalonsoportillo committed:
Implement code changes to enhance functionality and improve performance

nachoalonsoportillo committed:
fix: update example references in PrivateEndpointConnections and Servers JSON files; add VirtualNetworkSubnetUsageList example

nachoalonsoportillo committed:
fix: correct indentation in examples for long term retention and tuning options

nachoalonsoportillo committed:
Delete outdated example for creating PostgreSQL servers with Microsoft Entra enabled and update location references from 'westus' to 'eastus' in various examples. Add new examples for creating servers in Microsoft-owned and user-owned virtual networks with specific configurations, including high availability settings.

nachoalonsoportillo committed:
fix: update reference for example in Servers.json to correct file path

nachoalonsoportillo committed:
fix: update reference for example in Servers.json to correct file path

nachoalonsoportillo committed:
fix: update response structure in AdministratorsMicrosoftEntraAdd.json to reflect correct status codes and headers

nachoalonsoportillo committed:
fix: update example references and tags for server operations in Servers.json

nachoalonsoportillo committed:
fix: update response structure in AdministratorsMicrosoftEntraDelete.json to reflect correct status codes and headers

nachoalonsoportillo committed:
fix: update example references in Servers.json to correct file paths for start and stop operations

nachoalonsoportillo committed:
Refactor PostgreSQL examples to update async operation URLs and remove unnecessary response bodies

- Updated async operation URLs in ServersUpdateWithMicrosoftEntraEnabled.json, VirtualEndpointCreate.json, VirtualEndpointDelete.json, and VirtualEndpointUpdate.json to reflect the correct region and operation IDs.
- Removed redundant response bodies for 200 and 201 status codes in VirtualEndpointCreate.json and VirtualEndpointUpdate.json.
- Cleaned up response structure in VirtualEndpointDelete.json by removing unnecessary content field.

nachoalonsoportillo committed:
Refactor PostgreSQL examples for server promotion and tuning options

- Removed detailed response body for 200 status in examples of ServersPromoteReplicaAsForcedSwitchover, ServersPromoteReplicaAsPlannedStandaloneServer, and ServersPromoteReplicaAsPlannedSwitchover.
- Updated 202 status headers to use consistent "azure-asyncoperation" key and modified URLs for async operations.
- Added "kind" property to index recommendations in TuningOptionsListIndexRecommendations and TuningOptionsListIndexRecommendationsFilteredForCreateIndex examples.

nachoalonsoportillo committed:
fix: remove redundant 200 response schema from VirtualEndpoints.json

nachoalonsoportillo committed:
fix: update response headers and remove redundant response bodies in PrivateEndpointConnections examples

nachoalonsoportillo committed:
fix: streamline response structure and enhance details in TuningOptionsListTableRecommendationsFilteredForAnalyzeTable example

nachoalonsoportillo committed:
fix: update examples and descriptions to remove references to flexible servers

nasc17 commented:
Please address: 
<img width="626" height="115" alt="image" src="https://github.com/user-attachments/assets/3ee43893-4e42-4699-bedd-06c71533fd65" />

https://github.com/Azure/azure-rest-api-specs/pull/35537/checks?check_run_id=45819133544

@nachoalonsoportillo @mattboentoro

nasc17 commented:
@nachoalonsoportillo The changes in responses are showing many failures: https://github.com/Azure/azure-rest-api-specs/actions/runs/16226379147

We need to see if this can be addressed, suppressed, or reverted

nachoalonsoportillo committed:
Update Azure Async Operation headers in PostgreSQL examples for consistency and clarity

- Changed "azure-asyncoperation" to "Azure-AsyncOperation" in multiple JSON example files.
- Added "Location" header to responses in the following examples:
  - ServersUpdateWithMajorVersionUpgrade.json
  - ServersUpdateWithMicrosoftEntraEnabled.json
  - VirtualEndpointCreate.json
  - VirtualEndpointDelete.json
  - VirtualEndpointUpdate.json

nachoalonsoportillo committed:
fix: enhance database name description for clarity and detail

nachoalonsoportillo committed:
fix: add Azure-AsyncOperation header to multiple JSON specifications for consistency

Commit:
fix prettier

nachoalonsoportillo committed:
fix: add response headers for Azure-AsyncOperation in backup API

Commit:
fix issue with the closing bracket

nachoalonsoportillo committed:
fix: remove redundant response headers for 202 status in backups API

nachoalonsoportillo committed:
fix: update location references from westus to eastus in migration examples and specifications

nachoalonsoportillo committed:
fix: update migration examples and cancel operation details in PostgreSQL API

nachoalonsoportillo commented:
/azp run

nachoalonsoportillo committed:
fix: add example reference for updating existing migrations in PostgreSQL API

nachoalonsoportillo committed:
fix: update administrator login examples to use 'exampleadministratorlogin' across multiple JSON files

nachoalonsoportillo committed:
fix: add 204 No Content response description to multiple PostgreSQL API specifications

nachoalonsoportillo committed:
fix: add 204 No Content response to multiple PostgreSQL API example files

nachoalonsoportillo committed:
fix: update descriptions for CapabilityList, Migration, and Server to provide clarity

nachoalonsoportillo committed:
fix: add 204 No Content response to PrivateEndpointConnectionsDelete example

nachoalonsoportillo committed:
fix: rename CapabilityListResult and QuotaUsagesListResult to CapabilityList and QuotaUsageList for consistency

nachoalonsoportillo committed:
fix: rename enums for clarity and consistency across PostgreSQL specifications

nachoalonsoportillo committed:
fix: update descriptions in JSON files for consistency in quotation marks

nachoalonsoportillo committed:
fix: update descriptions in Migrations.json for clarity and consistency

nachoalonsoportillo committed:
fix: remove 'create' mutability from cluster properties in Servers.json

nachoalonsoportillo committed:
fix: correct casing in MigrationSubStateDetails reference for consistency

nachoalonsoportillo committed:
Refactor code structure for improved readability and maintainability

nachoalonsoportillo committed:
fix: correct casing in MigrationSubstateDetails for consistency

nachoalonsoportillo committed:
fix: correct casing in MigrationSubstate for consistency

nachoalonsoportillo committed:
fix: update suppression reasons for consistency in PostgreSQL namespace

nachoalonsoportillo committed:
fix: update suppression reasons for consistency in AdvancedThreatProtectionSettings and other JSON files

nachoalonsoportillo committed:
fix: update suppression reasons for consistency in PostgreSQL API contracts

nachoalonsoportillo committed:
fix: update suppression reasons and codes for consistency in PostgreSQL API contracts

nachoalonsoportillo committed:
fix: add suppression reasons for PatchResponseCodes in PostgreSQL API contracts

nachoalonsoportillo committed:
fix: add suppression for DeleteResponseBody in Migrations.json for API contract consistency

nachoalonsoportillo committed:
fix: update suppression code for DeleteResponseBody in Migrations.json for API contract consistency

nachoalonsoportillo committed:
fix: improve descriptions and add enum values for PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update migration references and example parameters for consistency in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update NameAvailability references and add description for server properties in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: remove outdated NameAvailability example files and add new examples for PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update AdvancedThreatProtectionSettings references and rename examples for consistency in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update parameters structure in NameAvailability examples for consistency in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update VirtualNetworkSubnetUsage references for consistency in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update OperationList reference and remove unused definitions in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update parameter references and remove unused parameters in Migrations.json for PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update parameter references for consistency in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update parameter references for consistency in Migrations.json for PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update targetDbServerName to serverName for consistency in PostgreSQL API specifications

nachoalonsoportillo committed:
fix: update targetDbServerName to serverName for consistency in Migrations.json for PostgreSQL API specifications

nachoalonsoportillo committed:
fix: add x-ms-pageable definition for pagination support in Operations.json

nachoalonsoportillo committed:
fix: update minorVersion to 5 in multiple example JSON files for PostgreSQL API specifications

nachoalonsoportillo committed:
fix: add detailed operation definitions and update reference in Operations.json for PostgreSQL API specifications

nachoalonsoportillo committed:
fix: rename additionalProperties to extraProperties for clarity in JSON specifications

nachoalonsoportillo committed:
fix: replace extraProperties with additionalProperties for consistency in JSON specifications

nachoalonsoportillo committed:
fix: update operation and server definitions for consistency in API specifications

nachoalonsoportillo committed:
fix: update references in Servers.json for patch operations and add new definitions for Backup, AuthConfig, and MaintenanceWindow

nachoalonsoportillo committed:
fix: update HighAvailability reference and add HighAvailabilityForPatch definition in Servers.json

nachoalonsoportillo committed:
fix: remove empty object definitions for 204 responses in JSON examples

nachoalonsoportillo committed:
Rename folder

nachoalonsoportillo committed:
Base commit

nachoalonsoportillo committed:
To facilitate comparison

nachoalonsoportillo committed:
fix: update reference to ServerList in Replicas.json to point to FlexibleServers.json

nachoalonsoportillo committed:
Add VirtualNetworkSubnetUsage API definition for PostgreSQL management

- Introduced a new API definition for checking virtual network subnet usage in Azure Database for PostgreSQL.
- Added paths, parameters, and response models to support the functionality.
- Defined necessary request and response schemas, including VirtualNetworkSubnetUsageParameter and VirtualNetworkSubnetUsageModel.
- Included delegated subnet usage details in the response model.

nachoalonsoportillo committed:
feat: add PUT operation for updating server configurations and example usage

nachoalonsoportillo committed:
fix: update suppression codes in readme.md to align with existing API contract

nachoalonsoportillo committed:
fix: update suppression code from PatchResponseCodes to PutResponseCodes in readme.md to align with existing API contract

nachoalonsoportillo committed:
Merge branch 'main' into mattboentoro/2025-06-01-preview
